rpcs3/Emu/Memory/vm.cpp

```diff
@@ -604,6 +613,14 @@ namespace vm
 			g_addr_lock.release(0);
 		}
 
+		for (u32 i = addr / 4096; i < addr / 4096 + size / 4096; i++)
+		{
+			if (!(g_pages[i].flags.exchange(0) & page_allocated))
+			{
+				fmt::throw_exception("Concurrent access (addr=0x%x, size=0x%x, current_addr=0x%x)" HERE, addr, size, i * 4096);
+			}
+		}
```

Comment:
I think that replacing 3 divisions with 1 should improve performance.
I don't think, I calculated everything correctly, so please check.)

```suggestion
		for (u32 i = addr; i < addr + size; i += 4096)
		{
			if (!(g_pages[i / 4096].flags.exchange(0) & page_allocated))
			{
				fmt::throw_exception("Concurrent access (addr=0x%x, size=0x%x, current_addr=0x%x)" HERE, addr, size, i);
			}
		}
```

Reply:
Thanks for the advice but what's really performance expensive in this loop is line 614 with `exchange(0)`, and you can't add `addr + size` like that because of potential overflow.

Reply:
Tou avoid overflow you need to substruct 1 -> `for (u32 i = addr, end = addr + size - 1; i < end; i += 4096)`

Reply:
I've pushed an optimized version.